File: LCS/compare/Command.cpp
// Copyright (C) 2017, Christopher N. Hume.  All rights reserved.
//
// 2017-06-29 CNHume  Created file
//
#include "Command.h"

string Command::arg_msg(const string& s, const string& name) {
  auto msg = string("Invalid ");
  msg += name;
  msg += ": ";
  msg += s;
  return msg;
}

uint32_t Command::parse_uint32(const char* s, const string& name) {
  char* pEnd;
  auto result = strtoul(s, &pEnd, 10);
  if (errno != 0 || *pEnd != '\0')
    throw invalid_argument(arg_msg(s, name));
  return result;
}

//
// command line parser
//
void Command::Parse(int argc, char* argv[]) {
  auto usage = false;
  auto n = 1;                           // skip verb
  for (; n < argc && !usage; n++) {     // parse any switches
    char* token = argv[n];
    auto len = strlen(token);
    if (len < 1 || token[0] != '-')     // end of switches
      break;
    else if (len < 2)
      usage = true;
    else {                              // parse single character switch
      switch (token[1]) {
      case 'b':                         // the ignorespace switch
        if (len > 2)                    // superfluous value specified
          usage = true;
        else
          ignorespace = true;
        break;

      case 'i':                         // the ignorecase switch
        if (len > 2)                    // superfluous value specified
          usage = true;
        else
          ignorecase = true;
        break;

      case 'j':                         // the join switch
        isjoin = true;
        if (len > 2)                    // whitespace optional
          join = parse_uint32(&token[2], "join");
        else if (n < argc)              // whitespace allowed
          join = parse_uint32(argv[++n], "join");
        else
          usage = true;
        break;

      case 'p':                         // the prefix switch
        if (len > 2)                    // whitespace optional
          prefix = parse_uint32(&token[2], "prefix");
        else if (n < argc)              // whitespace allowed
          prefix = parse_uint32(argv[++n], "prefix");
        else
          usage = true;
        break;

      case 's':                         // the suffix switch
        if (len > 2)                    // whitespace optional
          suffix = parse_uint32(&token[2], "suffix");
        else if (n < argc)              // whitespace allowed
          suffix = parse_uint32(argv[++n], "suffix");
        else
          usage = true;
        break;

      default:                          // switch unknown
        usage = true;
        break;
      }
    }
  }

  if (n < argc)                         // parse f1
    f1 = argv[n++];
  else                                  // f1 is required
    usage = true;

  if (n < argc)                         // parse f2
    f2 = argv[n++];
  else                                  // f2 is required
    usage = true;

  if (n < argc)                         // superfluous argument specified
    usage = true;

  if (usage)                            // throw usage line if parse failed
    throw runtime_error("Usage: compare [-b] [-i] [-j <join>] [-p <prefix>] [-s <suffix>] file1 file2");
}

File: C#/Searches/Anagram/Command.cs
//
// Copyright (C) 2018-2025, Christopher N. Hume.  All rights reserved.
//
// You should have received a copy of the MIT License along with this program.
// If not, see https://opensource.org/licenses/MIT.
//
// 2018-12-23 CNHume  Added Command class
//

namespace Parsers;

public class Command {
  #region Properties
  public string? Letters { get; set; }
  public bool Prefix { get; set; }
  public bool Sort { get; set; }
  #endregion

  #region Methods
  public void Parse(string[] args) {
    Letters = default;
    Prefix = false;

    var usage = false;
    var count = args.Length;
    var n = 0;
    for (; n < count && !usage; n++) {
      var token = args[n];              // parse any switches
      var len = token.Length;
      if (len < 1 || token[0] != '-')
        break;                          // end of switches
      else if (len < 2)
        usage = true;
      else {                            // parse single character switch
        switch (token[1]) {
        case 'p':                       // the prefix switch
          if (len > 2)                  // superfluous value specified
            usage = true;
          else
            Prefix = true;
          break;

        case 's':                       // the sort switch
          if (len > 2)                  // superfluous value specified
            usage = true;
          else
            Sort = true;
          break;

        default:                        // switch unknown
          usage = true;
          break;
        }
      }
    }

    // Letters are required
    if (n < count)
      Letters = args[n++];
    else
      usage = true;

    usage |= n < count;                 // superfluous argument specified

    if (usage)                        // throw usage line if parse failed
      throw new ApplicationException("Usage: Anagram [-p] [-s] letters");
  }
  #endregion
}
